pir/cpp/string_encoder.cpp

```diff
@@ -17,60 +17,131 @@
 }
 
-Status StringEncoder::encode(const string& value,
-                             Plaintext& destination) const {
-  size_t num_coeff =
-      ceil(static_cast<double>(value.size() * 8) / bits_per_coeff_);
+StatusOr<size_t> StringEncoder::calc_num_coeff(size_t num_bytes) const {
+  size_t num_coeff = ceil(static_cast<double>(num_bytes * 8) / bits_per_coeff_);
   if (num_coeff > poly_modulus_degree_) {
     return InvalidArgumentError(
-        "Number of coefficients needed to encode string greater than poly "
-        "modulus degree");
+        "Number of coefficients needed greater than poly modulus degree");
   }
+  return num_coeff;
+}
 
+Status StringEncoder::encode(const string& value,
+                             Plaintext& destination) const {
+  ASSIGN_OR_RETURN(const auto num_coeff, calc_num_coeff(value.size()));
   destination.resize(num_coeff);
   destination.set_zero();
-  size_t coeff_index = 0;
-  size_t coeff_bits = bits_per_coeff_;
-  for (uint8_t c : value) {
-    size_t remain_bits = 8;
-    while (remain_bits > 0) {
-      size_t n = std::min(coeff_bits, remain_bits);
-      destination[coeff_index] <<= n;
-      destination[coeff_index] |= (c >> (8 - n));
-      c <<= n;
-      coeff_bits -= n;
-      remain_bits -= n;
-      if (coeff_bits <= 0) {
-        ++coeff_index;
-        coeff_bits = bits_per_coeff_;
-      }
-    }
-  }
-  if (coeff_bits < bits_per_coeff_ && coeff_bits > 0) {
-    destination[coeff_index] <<= coeff_bits;
+  StringEncoderImpl impl(destination, bits_per_coeff_);
+  impl.encode(value);
+  impl.terminate();
+  return Status::OK;
+}
+
+Status StringEncoder::encode(vector<string>::const_iterator v,
+                             const vector<string>::const_iterator end,
+                             Plaintext& destination) const {
+  size_t total_size = std::accumulate(
+      v, end, 0, [](int a, const string& b) { return a + b.size(); });
+  ASSIGN_OR_RETURN(auto num_coeff, calc_num_coeff(total_size));
+  destination.resize(num_coeff);
+  destination.set_zero();
+  StringEncoderImpl impl(destination, bits_per_coeff_);
+  while (v != end) {
+    impl.encode(*(v++));
   }
+  impl.terminate();
   return Status::OK;
 }
 
-string StringEncoder::decode(const Plaintext& pt) const {
-  const size_t num_bytes = pt.significant_coeff_count() * bits_per_coeff_ / 8;
-  string result(num_bytes, 0);
+StatusOr<string> StringEncoder::decode(const Plaintext& pt, size_t length,
+                                       size_t byte_offset) const {
+  if ((byte_offset + length) > (pt.coeff_count() * bits_per_coeff_ / 8)) {
+    return InvalidArgumentError(
+        "Requested decode beyond end of data in polynomial");
+  }
+  size_t start_coeff_index = byte_offset * 8 / bits_per_coeff_;
+  size_t coeff_bits =
+      ((start_coeff_index + 1) * bits_per_coeff_) - (byte_offset * 8);
+  if (coeff_bits <= 0) {
+    // ++start_coeff_index;
+    coeff_bits = bits_per_coeff_;
+  }
+  if (length <= 0) {
```

Comment:
Same here, length is unsigned

Reply:
Again, I like coding as though it could go negative, just in case it changes later.